lib/pacemaker/pcmk_ticket.c

```diff
@@ -360,6 +389,93 @@ pcmk_ticket_info(xmlNodePtr *xml, const char *ticket_id)
     return rc;
 }
 
+int
+pcmk__ticket_remove_attr(pcmk__output_t *out, cib_t *cib, pcmk_scheduler_t *scheduler,
+                         const char *ticket_id, GList *attr_delete)
+{
+    xmlNode *ticket_state_xml = NULL;
+    xmlNode *xml_top = NULL;
+    int rc = pcmk_rc_ok;
+
+    CRM_ASSERT(out != NULL && cib != NULL && scheduler != NULL);
+
+    if (ticket_id == NULL) {
+        return EINVAL;
+    }
+
+    /* Nothing to do */
+    if (attr_delete == NULL || g_list_length(attr_delete) == 0) {
+        return pcmk_rc_ok;
+    }
+
+    rc = build_ticket_modify_xml(cib, ticket_id, &ticket_state_xml, &xml_top);
+
+    if (rc == pcmk_rc_duplicate_id) {
+        out->info(out, "Multiple " PCMK__XE_TICKET_STATE "s match ticket=%s", ticket_id);
+    } else if (rc != pcmk_rc_ok) {
+        free_xml(ticket_state_xml);
+        return rc;
+    }
+
+    for (GList *list_iter = attr_delete; list_iter != NULL; list_iter = list_iter->next) {
+        const char *key = list_iter->data;
+        pcmk__xe_remove_attr(ticket_state_xml, key);
+    }
+
+    crm_log_xml_debug(xml_top, "Replace");
+    rc = cib->cmds->replace(cib, PCMK_XE_STATUS, ticket_state_xml, cib_sync_call);
+    rc = pcmk_legacy2rc(rc);
+
+    free_xml(xml_top);
+    return rc;
+}
+
+int
+pcmk_ticket_remove_attr(xmlNodePtr *xml, const char *ticket_id, GList *attr_delete)
+{
+    pcmk_scheduler_t *scheduler = NULL;
+    pcmk__output_t *out = NULL;
+    int rc = pcmk_rc_ok;
+    cib_t *cib = NULL;
+
+    rc = pcmk__xml_output_new(&out, xml);
+    if (rc != pcmk_rc_ok) {
+        return rc;
+    }
+
+    cib = cib_new();
+    if (cib == NULL) {
+        rc = pcmk_rc_cib_corrupt;
+        goto done;
+    }
+
+    rc = cib->cmds->signon(cib, crm_system_name, cib_command);
+    rc = pcmk_legacy2rc(rc);
+
+    if (rc != pcmk_rc_ok) {
+        goto done;
+    }
+
+    rc = pcmk__init_scheduler(out, NULL, NULL, &scheduler);
+    if (rc != pcmk_rc_ok) {
+        goto done;
+    }
+
+    pcmk__unpack_constraints(scheduler);
+
+    pcmk__register_lib_messages(out);
```

Comment:
could functionize everything before the private API call

Reply:
Yeah, I think there's a couple variations on this pattern that could be turned into functions.  I got a little stumped thinking about how wide spread they could be used in libpacemaker, and where they would live, etc. and then just moved on.  I think I'll functionize things just in this file and we can always move them around later as needed.